channel/src/main/java/com/synopsys/integration/alert/channel/jira/common/JiraCustomFieldResolver.java

```diff
@@ -0,0 +1,112 @@
+        this.retrieveAvailableFields = retrieveAvailableFields;
+        this.fieldCache = new HashMap<>();
+        this.isCachePopulated = false;
+    }
+
+    public final JiraResolvedCustomField resolveCustomField(JiraCustomFieldConfig jiraCustomFieldConfig) {
+        CustomFieldDefinitionModel fieldDefinition = retrieveCustomFieldDefinition(jiraCustomFieldConfig);
+        Object requestObject = convertValueToRequestObject(fieldDefinition, jiraCustomFieldConfig);
+        return new JiraResolvedCustomField(fieldDefinition.getFieldId(), requestObject);
+    }
+
+    protected Optional<CustomFieldCreationResponseModel> retrieveFieldDefinition(String fieldName) {
+        if (!isCachePopulated) {
+            try {
+                initializeCache();
+            } catch (IntegrationException e) {
+                logger.warn("No Jira Cloud user-visible fields found");
+                return Optional.empty();
+            }
+        }
+        return Optional.ofNullable(fieldCache.get(fieldName));
+    }
+
+    protected CustomFieldDefinitionModel retrieveCustomFieldDefinition(JiraCustomFieldConfig customFieldConfig) {
+        String fieldName = customFieldConfig.getFieldName();
+        CustomFieldCreationResponseModel fieldResponse = retrieveFieldDefinition(fieldName)
+                                                             .orElseThrow(() -> new AlertRuntimeException(String.format("No custom field named '%s' existed", fieldName)));
+        return new CustomFieldDefinitionModel(fieldResponse.getId(), fieldResponse.getSchema().getType());
+    }
+
+    protected Object convertValueToRequestObject(CustomFieldDefinitionModel fieldDefinition, JiraCustomFieldConfig jiraCustomFieldConfig) {
+        String fieldType = fieldDefinition.getFieldType();
+        switch (fieldType) {
+            case CUSTOM_FIELD_TYPE_STRING_VALUE:
+                return jiraCustomFieldConfig;
+            case CUSTOM_FIELD_TYPE_ARRAY_VALUE:
+                JsonArray jsonArray = new JsonArray();
+                jsonArray.add(jiraCustomFieldConfig.getFieldValue());
+                return jsonArray;
+            case CUSTOM_FIELD_TYPE_OPTION_VALUE:
+                JsonObject jsonObject = new JsonObject();
+                jsonObject.addProperty("value", jiraCustomFieldConfig.getFieldValue());
+                return jsonObject;
+            default:
+                throw new AlertRuntimeException(String.format("Unsupported field type '%s' for field: %s", fieldType, jiraCustomFieldConfig.getFieldName()));
+        }
+    }
+
+    private void initializeCache() throws IntegrationException {
+        List<CustomFieldCreationResponseModel> userVisibleFields = retrieveAvailableFields.get();
+        for (CustomFieldCreationResponseModel fieldModel : userVisibleFields) {
+            fieldCache.put(fieldModel.getName(), fieldModel);
+        }
+        isCachePopulated = true;
+    }
```

Comment:
How long are we keeping this cache around for?

Reply:
For the duration of processing for a single event.